glean-core/src/event_database/mod.rs

```diff
@@ -0,0 +1,325 @@
+    pub fn serialize_relative(&self, timestamp_offset: u64) -> JsonValue {
+        json!(&RecordedEventData {
+            timestamp: self.timestamp - timestamp_offset,
+            category: self.category.clone(),
+            name: self.name.clone(),
+            extra: self.extra.clone(),
+        })
+    }
+}
+
+/// This struct handles the in-memory and on-disk storage logic for events.
+///
+/// So that the data survives shutting down of the application, events are stored
+/// in an append-only file on disk, in addition to the store in memory. Each line
+/// of this file records a single event in JSON, exactly as it will be sent in the
+/// ping. There is one file per store.
+///
+/// When restarting the application, these on-disk files are checked, and if any are
+/// found, they are loaded, queued for sending and flushed immediately before any
+/// further events are collected. This is because the timestamps for these events
+/// may have come from a previous boot of the device, and therefore will not be
+/// compatible with any newly-collected events.
+#[derive(Debug)]
+pub struct EventDatabase {
+    pub path: PathBuf,
+    // The in-memory list of events
+    event_stores: RwLock<HashMap<String, Vec<RecordedEventData>>>,
+    file_lock: RwLock<()>,
+}
+
+impl EventDatabase {
+    /// Create a new event database.
+    ///
+    /// # Arguments
+    ///
+    /// * `data_path` - The directory to store events in. A new directory
+    ///   `events` will be created inside of this directory.
+    pub fn new(data_path: &str) -> Result<Self> {
+        let path = Path::new(data_path).join("events");
+        create_dir_all(&path)?;
+
+        Ok(Self {
+            path,
+            event_stores: RwLock::new(HashMap::new()),
+            file_lock: RwLock::new(()),
+        })
+    }
+
+    /// Initialize events storage. This must be called once on application startup,
+    /// e.g. from [Glean.initialize], but after we are ready to send pings, since
+    /// this could potentially collect and send pings.
+    ///
+    /// If there are any events queued on disk, it loads them into memory so
+    /// that the memory and disk representations are in sync.
+    ///
+    /// Secondly, if this is the first time the application has been run since
+    /// rebooting, any pings containing events are assembled into pings and cleared
+    /// immediately, since their timestamps won't be compatible with the timestamps
+    /// we would create during this boot of the device.
+    ///
+    /// # Arguments
+    ///
+    /// * `glean` - The Glean instance.
+    pub fn on_ready_to_send_pings(&self, glean: &Glean) {
+        match self.load_events_from_disk() {
+            Ok(_) => self.send_all_events(glean),
+            Err(err) => log::error!("Error loading pings from disk: {}", err),
+        }
+    }
+
+    fn load_events_from_disk(&self) -> Result<()> {
+        let _lock = self.file_lock.read().unwrap();
```

Comment:
What are we locking here?
The nice thing about locks in Rust that they usually wrap the data that ought to be protected.
Now that we have nothing in here, it's not clear what's protected.

Reply:
We're locking for operations on the filesystem.  I just added a comment to this field in the struct to clarify this.

In this case, we don't want to read in events from files when we might also be deleting them in snapshot_as_json.  Since Android is Linux, the files themselves have locking semantics.  What doesn't work is iterating over a directory while deleting those files in some other thread.

In practice right now, this shouldn't happen given how we queue work on the Kotlin side, but if that were to change we could be in trouble, and since these functions don't have a mutable reference to the EventDatabase instance, I don't think Rust would enforce the synchronization.

All of this comes from actual problems we had with intermittent tests in glean-ac...

Reply:
Maybe during backlog/tech debt cleanup we can take another look if we can solve this differently. For now I don't worry too much.